apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.js

```diff
@@ -81,256 +91,197 @@ router.post(
     }
-    if (res.locals.question.qid === qid_new) {
-      debug('The new qid is the same as the old qid - do nothing');
-      res.redirect(req.originalUrl);
-    } else {
-      const editor = new QuestionRenameEditor({
-        locals: res.locals,
-        qid_new: qid_new,
-      });
-      editor.canEdit((err) => {
-        if (ERR(err, next)) return;
-        editor.doEdit((err, job_sequence_id) => {
-          if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-            res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
-          } else {
-            res.redirect(req.originalUrl);
-          }
+    if (req.body.__action === 'change_id') {
+      if (!req.body.id) throw error.make(400, `Invalid QID (was falsy): ${req.body.id}`);
+      if (!/^[-A-Za-z0-9_/]+$/.test(req.body.id)) {
+        throw error.make(
+          400,
+          `Invalid QID (was not only letters, numbers, dashes, slashes, and underscores, with no spaces): ${req.body.id}`,
+        );
+      }
+      let qid_new;
+      try {
+        qid_new = path.normalize(req.body.id);
+      } catch (err) {
+        throw error.make(400, `Invalid QID (could not be normalized): ${req.body.id}`);
+      }
+      if (res.locals.question.qid === qid_new) {
+        res.redirect(req.originalUrl);
+      } else {
+        const editor = new QuestionRenameEditor({
+          locals: res.locals,
+          qid_new: qid_new,
         });
-      });
-    }
-  } else if (req.body.__action === 'copy_question') {
-    debug('Copy question');
-    if (idsEqual(req.body.to_course_id, res.locals.course.id)) {
-      // In this case, we are making a duplicate of this question in the same course
-      const editor = new QuestionCopyEditor({
-        locals: res.locals,
-      });
-      editor.canEdit((err) => {
-        if (ERR(err, next)) return;
-        editor.doEdit((err, job_sequence_id) => {
-          if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-            res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
-          } else {
-            debug(
-              `Get question_id from uuid=${editor.uuid} with course_id=${res.locals.course.id}`,
-            );
-            sqldb.queryOneRow(
-              sql.select_question_id_from_uuid,
-              { uuid: editor.uuid, course_id: res.locals.course.id },
-              (err, result) => {
-                if (ERR(err, next)) return;
-                flash(
-                  'success',
-                  'Question copied successfully. You are now viewing your copy of the question.',
-                );
-                res.redirect(
-                  res.locals.urlPrefix + '/question/' + result.rows[0].question_id + '/settings',
-                );
-              },
-            );
-          }
+        editor.assertCanEdit();
+        const serverJob = await editor.setupEditAsync();
+        try {
+          await editor.doEditAsync(serverJob);
```

Comment:
As long as we're redesigning this API, here are a few thoughts:

- Always call `assertCanEdit` in `setupEditAsync`. I did a skim of the codebase and we always have `assertCanEdit`/`canEdit` immediately followed by `doEdit`/`setupEditAsync`. If that's the case, let's save the caller some trouble and just always do the validation there.
- Bikeshedding on names: how about `prepareServerJob` and `executeWithServerJob` instead of `setupEditAsync` and `doEditAsync`?

Reply:
> Always call assertCanEdit in setupEditAsync. I did a skim of the codebase and we always have assertCanEdit/canEdit immediately followed by doEdit/setupEditAsync. If that's the case, let's save the caller some trouble and just always do the validation there.

That makes sense. I have now added a call for `assertCanEdit()` from inside `setupEditAsync` and removed `assertCanEdit()` from `instructorQuestionSettings.ts`. Should we update other calls of `canEdit()` throughout the codebase as part of this PR or separately? 

>Bikeshedding on names: how about prepareServerJob and executeWithServerJob instead of setupEditAsync and doEditAsync?

I like `prepareServerJob` and `executeWithServerJob`. I have updated to these names in [ea3d984](https://github.com/PrairieLearn/PrairieLearn/pull/9439/commits/ea3d984b36cb388f772a3f33dcda7bc4224a507c). Do we need to update the name of the callback function (`doEdit`) to be more similar to these?

Reply:
> Should we update other calls of canEdit() throughout the codebase as part of this PR or separately?

Let's do it separately; that will also require changing from callbacks to async/await, which is IMO too invasive of a change for this particular PR.

> Do we need to update the name of the callback function (doEdit) to be more similar to these?

No, I don't think so. The general pattern we use for things like this is to introduce a new API, gradually replace all usages of the old API with the new one, and then delete the old API when it's no longer needed. We do this all without changing the new API to save ourselves the trouble of having to change things twice. Does that make sense?